### PR TITLE
Improve `RecordStreamAccess` and `EventualRecordStreamAssertion`

### PR DESCRIPTION
**Pull request by tinker-michaelj**
**Description**:
 - Updates `RecordStreamAccess` to support a watcher pattern and provide efficient "real-time" access to `RecordStreamItem`'s.
 - Creates an `EventualRecordStreamAssertion` util op that, _given an assertion about the record stream that should pass within a timeout_, subscribes to `RecordStreamAccess`; and then continually updates its assertion status based on the stream.
 - Updates `HapiSpec` to give special treatment to operations of type `EventualRecordStreamAssertion`, ensuring they get steady background traffic to close record stream files; and blocking on their assertions at the end of the spec.

### CODE REVIEW

**Review by iwsimon — DISMISSED** [review]
LGTM

**Review by mhess-swl — COMMENTED** [review]
It took me a bit to get my head around this, but now that I get the idea it's really cool! Well done!

File: test-clients/src/main/java/com/hedera/services/bdd/junit/RecordStreamAccess.java [hunk 1]
```diff
@@ -29,9 +31,74 @@
...
+
+    /**
+     * If the given location is not already being watched, starts a new listener for it and returns
+     * the listener.
+     *
+     * @param loc the record stream file location to watch
+     * @return the listener for the given location
+     * @throws Exception if there is an error starting the listener
+     */
+    public synchronized BroadcastingRecordStreamListener getValidatingListener(final String loc)
```
[the remark this answers is not in the document]
Comment (tinker-michaelj) [reply test-clients/src/main/java/com/hedera/services/bdd/junit/RecordStreamAccess.java:85]: True, protecting against the scenario that concurrent `HapiSpec`'s `A` and `B` interleave as follows:

1. `A` calls `stopMonitorIfNoSubscribers()`, which computes `numSubscribers = 0`.
2. `B` calls `getValidatingListener()` and registers a listener that needs a running monitor.
3. Now `A` runs again and calls `monitor.stop()`.

File: test-clients/src/main/java/com/hedera/services/bdd/suites/crypto/NewAccountRecordExists.java [hunk 2]
```diff
@@ -79,6 +83,23 @@ private HapiSpec newAccountIsReflectedInRecordStream() {
...
+        return onlyDefaultHapiSpec("NewAccountIsReflectedInRecordStreamV2")
+                .given(
+                        streamMustInclude(
+                                recordedCryptoCreate(
+                                        account, a -> a.withMemo(memo).withBalance(balance))))
+                .when(cryptoCreate(account).balance(balance).memo(memo))
+                .then(
+                        // HapiSpec automatically waits for the streamMustInclude()
+                        // assertion to pass, fail, or time out while ensuring enough
+                        // background traffic to keep closing record stream files
```
[the remark this answers is not in the document]
Comment (tinker-michaelj) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/crypto/NewAccountRecordExists.java:99]: Adding a `awaitStreamAssertions()` util verb.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/streams/assertions/CryptoCreateAssertion.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                + expectedMemo
+                + '\''
+                + ", expectedBalance="
+                + expectedBalance
+                + '}';
+    }
+
+    private long amountCredited(final TransferList transfers, final AccountID beneficiary) {
+        return transfers.getAccountAmountsList().stream()
+                .filter(aa -> aa.getAccountID().equals(beneficiary))
```
[the remark this answers is not in the document]
Comment (tinker-michaelj) [reply test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/streams/assertions/CryptoCreateAssertion.java:101]: `AccountAmount`

File: test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/streams/RecordStreamAssertion.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    /**
+     * Updates the assertion's state based on a relevant {@link RecordStreamItem}, throwing an
+     * {@link AssertionError} if a failure state is reached; or returning true if the assertion has
+     * reached a success state.
+     *
+     * @param item the item to test
+     * @throws AssertionError if the assertion has failed
+     * @return true if the assertion has succeeded
+     */
+    boolean updateAndTest(RecordStreamItem item) throws AssertionError;
```
[the remark this answers is not in the document]
Comment (tinker-michaelj) [reply test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/streams/RecordStreamAssertion.java:51]: Agreed, changed to `test()`.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/streams/RecordStreamAssertion.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+ * limitations under the License.
+ */
+package com.hedera.services.bdd.spec.utilops.streams;
+
+import com.hedera.services.bdd.spec.utilops.streams.assertions.CryptoCreateAssertion;
+import com.hedera.services.stream.proto.RecordStreamItem;
+
+/**
+ * Implements an assertion about one or more {@link RecordStreamItem}'s that should appear in the
+ * record stream during---or shortly after---execution of a {@link
```
[the remark this answers is not in the document]
Comment (tinker-michaelj) [reply test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/streams/RecordStreamAssertion.java:23]: I was just imagining a `HapiSpec` that ran long enough to close multiple record stream files.

In that case, one of the `EventualRecordStreamAssertion`'s might pass (on the `FileAlterationMonitor` thread) while other `HapiSpec` operations are still running on the "main" threads.

**Review by mhess-swl — APPROVED** [review]

**Review by mhess-swl — APPROVED** [review]

**Pull request merged**